src/java/org/apache/cassandra/net/BufferPoolAllocator.java

```diff
@@ -94,17 +98,37 @@ void release()
     public static class Wrapped extends UnpooledUnsafeDirectByteBuf
     {
         private ByteBuffer wrapped;
+        private boolean returnToPool = true;
 
         Wrapped(BufferPoolAllocator allocator, ByteBuffer wrap, int maxCapacity)
         {
             super(allocator, wrap, max(wrap.capacity(), maxCapacity));
             wrapped = wrap;
         }
 
+        @Override
+        public ByteBuf capacity(int newCapacity)
+        {
+            if (newCapacity == capacity())
+                return this;
+
+            // resizing doesn't use the pool
```

Comment:
I don't understand why the pool isn't used here. You mean you've checked the code for `UnpooledUnsafeDirectByteBuf`and it doesn't use the pool upon resizing?

Reply:
Initially, I noticed it doesn't use the pool because releasing a resized buffer made the pool's occupancy go negative.
In other words, calling `GlobalBufferPoolAllocator.instance.usedSizeInBytes()` after `resizedBuffer.release()` would return a negative value.

And I confirmed that by looking at netty's source too:

https://github.com/netty/netty/blob/netty-4.1.58.Final/buffer/src/main/java/io/netty/buffer/UnpooledDirectByteBuf.java#L156

https://github.com/netty/netty/blob/netty-4.1.58.Final/buffer/src/main/java/io/netty/buffer/UnpooledDirectByteBuf.java#L104

It calls `ByteBuffer.allocateDirect` and doesn't use the internal allocator we pass in the constructor, which explains the behavior I described above. Since we're extending **Unpooled**UnsafeDirectByteBuf, I guess it makes sense.

The check I added around `returnToPool` prevents putting a ByteBuffer back into the pool that hasn't originated in the pool.

Reply:
OK yes I was referring to that, whether you had checked we were keeping it consistent with netty.